src/Microsoft.AspNet.Mvc.TagHelpers/Internal/GlobbingUrlBuilder.cs

```diff
@@ -189,60 +207,172 @@ public int Compare(string x, string y)
                 yExtIndex = yExtIndex > ySlashIndex ? yExtIndex : -1;
 
                 // Get paths without their extensions, if they have one
-                var xNoExt = xExtIndex >= 0 ? x.Substring(0, xExtIndex) : x;
-                var yNoExt = yExtIndex >= 0 ? y.Substring(0, yExtIndex) : y;
+                var xLength = xExtIndex >= 0 ? xExtIndex : x.Length;
+                var yLength = yExtIndex >= 0 ? yExtIndex : y.Length;
+                var compareLength = Math.Max(xLength, yLength);
 
-                if (string.Equals(xNoExt, yNoExt, StringComparison.Ordinal))
+                if (string.Compare(x, 0, y, 0, compareLength, StringComparison.Ordinal) == 0)
                 {
                     // Only extension differs so just compare the extension
-                    var xExt = xExtIndex >= 0 ? x.Substring(xExtIndex) : string.Empty;
-                    var yExt = yExtIndex >= 0 ? y.Substring(yExtIndex) : string.Empty;
-                    return string.Compare(xExt, yExt, StringComparison.Ordinal);
+                    if (xExtIndex >= 0 && yExtIndex >= 0)
+                    {
+                        var length = x.Length - xExtIndex;
+                        return string.Compare(x, xExtIndex, y, yExtIndex, length, StringComparison.Ordinal);
+                    }
+
+                    return xExtIndex - yExtIndex;
                 }
 
-                var xSegments = xNoExt.Split(new[] { '/' }, StringSplitOptions.RemoveEmptyEntries);
-                var ySegments = yNoExt.Split(new[] { '/' }, StringSplitOptions.RemoveEmptyEntries);
+                var xNoExt = xExtIndex >= 0 ? x.Substring(0, xExtIndex) : x;
+                var yNoExt = yExtIndex >= 0 ? y.Substring(0, yExtIndex) : y;
 
-                if (xSegments.Length != ySegments.Length)
+                var result = 0;
+                var xEnumerator = new StringTokenizer(xNoExt, '/').GetEnumerator();
+                var yEnumerator = new StringTokenizer(yNoExt, '/').GetEnumerator();
+                StringSegment xSegment;
+                StringSegment ySegment;
+                while (TryGetNextSegment(ref xEnumerator, out xSegment))
                 {
-                    // Different path depths so shallower path wins
-                    return xSegments.Length.CompareTo(ySegments.Length);
+                    if (!TryGetNextSegment(ref yEnumerator, out ySegment))
```

Comment:
You're probably fixing a bug here e.g. wouldn't previous code place `/ZZ/a.txt` earlier than `/aa/bb/c.txt`? Please ensure you have tests covering such comparisons.